src/traces/scattergl/select.js

```diff
@@ -0,0 +1,70 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var subtypes = require('../scatter/subtypes');
+
+module.exports = function selectPoints(searchInfo, polygon) {
+    var cd = searchInfo.cd,
+        xa = searchInfo.xaxis,
+        ya = searchInfo.yaxis,
+        selection = [],
+        trace = cd[0].trace,
+        curveNumber = trace.index,
+        i,
+        di,
+        x,
+        y;
+
+    var scattergl = cd[0].glTrace;
+    var scene = cd[0].glTrace.scene;
+
+    // TODO: include lines? that would require per-segment line properties
+    var hasOnlyLines = (!subtypes.hasMarkers(trace) && !subtypes.hasText(trace));
+    if(trace.visible !== true || hasOnlyLines) return;
+
+    // filter out points by visible scatter ones
+    // var scatter2d = scattergl.scatter.instance
+
+    if(polygon === false) { // clear selection
+        for(i = 0; i < cd.length; i++) cd[i].dim = 0;
+    }
+    else {
+        for(i = 0; i < cd.length; i++) {
+            di = cd[i];
+            //FIXME: this affects performance for 1e6 points
+            x = xa.c2p(di.x);
+            y = ya.c2p(di.y);
```

Comment:
@alexcjohnson yes it's fairly simple as a CPU algo for a specific point, but the GPU needs to do it in parallel and it can't even do proper loops or index handling. The best analogy would be [ray marching](https://en.wikipedia.org/wiki/Volume_ray_casting) that underpins almost all [shadertoy](https://www.shadertoy.com/view/Xds3zN) creations, and it samples a 3D space populated with 3D shapes that are individually easy to describe mathematically (eg. modeling a head as a sphere and a neck as a cylinder, combining them via [CSG](https://en.wikipedia.org/wiki/Constructive_solid_geometry). Arbitrary 3D shapes can be built. The raymarching samples the 3D space by shooting rays as per the 2D pixel grid. 

It's possible to do a 2D variant of this; it'd require that the lassoed area is decomposed of mathematically easy to model primitives eg. triangles (so, just triangulation of the lassoed area). It uses [signed distance fields](http://iquilezles.org/www/articles/distfunctions/distfunctions.htm) with which there's ample experience on the team.

As it's not for artistic output but data analysis, I think significant time would go into a robust approach even though it's "only" 2D.